### PR TITLE
Modal: Add small top padding to the content so that avoid cutting off the visible outline when hovering items

### PR DESCRIPTION
**Pull request by t-hamano**
_Update:_

This PR originally tried to fix the cutoff when focusing on an item in the "Add template" modal. However, with #51806, the text now appears above the item and I have confirmed that this problem no longer occours.

However, the same problem occurs when there is a focusable element at the beginning of the modal content. Therefore, I decided to add 2px top padding to the content of the modal component itself.

## What?
This PR adds 2px top padding to the content of the `Modal` component. This prevents the problem of the outline being cut off in advance when a focusable element is inserted into the modal content.

modal_cutoff

## How?
Use the `$border-width-focus-fallback` variable to add 2px padding to the top of the modal content.

### Rename Navigation Modal

- Access Site Editor > Navigation > Click ellipsis button > Select "Rename".
- This PR removed the top padding of the rename modal form, but thanks to the top padding of the modal component itself, the focus outline should not cut off as before.

navigation-rename-modal

### Choose Pattern Modal

I have found that adding top padding to the Modal component itself fixes the outline cut-off in the following two existing modals:

- The Choose Pattern Modal displayed when creating a custom template
- The Choose Pattern Modal displayed when creating a new post

choose a pattern

These modals [already have 2px top padding to prevent outline cutting on focus](https://github.com/WordPress/gutenberg/blob/a3af2c7d0aee82a0cb8f74beae95730849b5afc6/packages/edit-site/src/components/start-template-options/style.scss#L32-L33). However, when focused, the outlines are [extended by an additional 2px due to the offset](https://github.com/WordPress/gutenberg/blob/a3af2c7d0aee82a0cb8f74beae95730849b5afc6/packages/block-editor/src/components/block-patterns-list/style.scss#L39-L41), resulting in the outlines being cut off. By adding top padding to the modal component itself, it now has a total of 4px top padding, which also fixes the outline cutoff on focus.

### CODE REVIEW

**Comment by t-hamano** [comment]
Since Gutenberg 16.1 has already been released, I have added the backport label just in case.

**Comment by jasmussen** [comment]
Relevant: #51806. CC: @jameskoster

**Comment by jasmussen** [comment]
Is there any way we can fix this in the modal component itself? I.e. right now there's a little bottom padding baked into the modal header:

Screenshot 2023-06-23 at 10 25 03

But there's no top padding baked into the content area:

Screenshot 2023-06-23 at 10 25 11

Could we split that padding in two, have half be bottom padding on the header, and the other half be top padding on the content?

**Comment by jameskoster** [comment]
If we split the padding then the stroke which appears beneath the header on scroll might look a bit close to the title text.

Adding 4px padding-top to `.components-modal__content` could be worth a try?

**Comment by t-hamano** [comment]
This problem will always occur when a focusable element, such as a button, textbox, etc., is placed at the beginning of the modal content. Therefore, it might be a good idea to add slight padding to the top of the modal content.

@ciampo

Is there any problem with the approach of adding a slight padding at the top of the `.components-modal__content` to prevent the cutoff of the focus outline?

modal_cutoff

**Comment by ciampo** [comment]
I understand the problem, but we shouldn't be using internal components classnames (like `.components-modal__content`) outside of the components package.

Couldn't we add small `margin-top` on the `edit-site-add-new-template__template-list__contents` grid container? Alternatively, could we change the `Modal` component to always have a small padding on its top container? In both cases, `2px` would be enough to avoid cutting off the focus ring

**Comment by t-hamano** [comment]
Thanks for the advice, @ciampo.

> Couldn't we add small margin-top on the edit-site-add-new-template__template-list__contents grid container? Alternatively, could we change the Modal component to always have a small padding on its top container? In both cases, 2px would be enough to avoid cutting off the focus ring

I would like to try the latter method so that we don't have to add `padding-top` every time we encounter similar problems in the future.

**Comment by t-hamano** [comment]
I have changed this PR as follows:

- Added 2px top padding to the content of the modal component itself. This should cause no outline cutoff when the element with the outline appears at the beginning of the content when in focus.
- The top padding of the navigation rename modal has been removed as it is no longer needed.

As a result, the target to be fixed by this PR has changed. Please see details in the description of the updated PR.

File: packages/components/src/modal/style.scss [hunk 1]
```diff
@@ -130,7 +130,7 @@
 .components-modal__content {
 	flex: 1;
 	margin-top: $header-height + $grid-unit-15;
-	padding: 0 $grid-unit-40 $grid-unit-40;
+	padding: $border-width-focus-fallback $grid-unit-40 $grid-unit-40;
```
Comment (ciampo) [inline packages/components/src/modal/style.scss]: It may be worth to add a short inline comment explaining the reason for the top padding

**Review by ciampo — APPROVED** [review]
LGTM, although I'll let Jay/Joen give the final green light

**Comment by jameskoster** [comment]
I would be tempted to use 4px, so that the overall dimensions continue to align with the 4px baseline grid.

4px would also account for more elaborate focus styles, such as pattern previews, which have an effective 3px outline (1.5px white + 1.5px admin theme color):

Screenshot 2023-06-26 at 09 38 22

**Comment by jasmussen** [comment]
This seems to solve the issue in a generic way. Top padding is the same:

Screenshot 2023-06-26 at 10 38 06

Content padding is now 2px at the top:

Screenshot 2023-06-26 at 10 37 54

However if I don't misread this, that means the spacing between the header title and the body content is now 2px taller, which puts the metrics off a little bit. 

So to simplify the whole thing, I would:

* Apply the 4px padding at the top of the content part (`$grid-unit-05`)
* _Subtract_ those same 4px from the bottom padding of the header 

So for the header instead of `padding: $grid-unit-30 $grid-unit-40 $grid-unit-15;`, you'd have something like `padding: ($grid-unit-30 - $grid-unit-05) $grid-unit-40 $grid-unit-15;`

**Comment by t-hamano** [comment]
@jasmussen

> * _Subtract_ those same 4px from the bottom padding of the header 

The `Modal` component draws a border at the bottom of the header if the content is scrollable. Therefore, reducing the bottom padding of the modal header by 4px will cause the header text to drop slightly

### Bottom padding is 12px

before

### Bottom padding is 8px

after

The approach of changing the top padding of the modal content from 2px to 4px seems to have less visual impact. What do you think?

**Comment by jasmussen** [comment]
It's a bit hard to discern the difference between those two screenshots, which to me suggests it's okay for that text to drop a bit. 

The default metrics should still ideally be somewhat close to this grid:
Screenshot 2023-06-26 at 15 40 15

Hope that helps!

**Comment by t-hamano** [comment]
I have made the following changes:

- Top padding of the modal content to **4px**
- Bottom padding of the modal header from **8px** to **4px**
- Top padding, no longer needed in the pattern selection modal, has also been removed

This looks fine to me. Below are some screenshots.

<details><summary>Preferences</summary>

preferences

</details> 

<details><summary>Keyboard shotycuts (Scrollable)</summary>

keyboard-shortcuts

</details> 


<details><summary>Rename Navigation</summary>

navigation-rename

</details> 


<details><summary>Choose Pattern Modal displayed when creating a custom template</summary>

choose-pattern-new-template

</details> 

<details><summary>Choose Pattern Modal displayed when creating a new post</summary>

choose-pattern-new-post

</details> 

<details><summary>Choose Pattern Modal displayed when creating a new post (mobile)</summary>

choose-pattern-new-post-mobile

</details> 


<details><summary>Rename Navigation (mobile)</summary>

rename-mobile

</details> 

<details><summary>ConfirmDialog (no header)</summary>

confirm-dialog

</details>

**Review by jasmussen — APPROVED** [review]
Thanks for the patience. Looks good:

Screenshot 2023-06-27 at 10 33 12
Screenshot 2023-06-27 at 10 33 16

The preferences dialog looks a little custom, and does not have the top padding, the tabs even misalign:
Screenshot 2023-06-27 at 10 32 37

However that's also in trunk, so separate:
Screenshot 2023-06-27 at 10 33 48

Nice work, thank you!

**Comment by tellthemachines** [comment]
I just cherry-picked this PR to the add/pre-beta1-bugfixes branch to get it included in the next release: 4f3599df90

**Comment by stokesman** [comment]
I think this caused a little visual regression on Blocks screen of the Edit Post Preferences Modal:

image

Gap above the sticky elements.

**Pull request merged**